File: src/optic/lua/flat/ops/Block.java
package optic.lua.flat.ops;

import optic.lua.flat.*;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Stream;

public class Block implements Step {
	private final List<Step> steps;
	private final StepType type;

	public Block(StepType type, List<Step> steps) {
		this.type = Objects.requireNonNull(type);
		this.steps = List.copyOf(steps);
	}

	@Override
	public @NotNull StepType getType() {
		return type;
	}

	@Override
	public @NotNull Stream<Step> children() {
		return steps.stream();
	}

	@Override
	public String toString() {
		return typeName();
	}
}

File: src/optic/lua/asm/instructions/Block.kt
package optic.lua.asm.instructions

import optic.lua.asm.AsmBlock
import optic.lua.asm.Step
import optic.lua.asm.StepVisitor

class Block(val steps: AsmBlock) : Step {
    override fun <T : Any, X : Throwable> accept(visitor: StepVisitor<T, X>): T = visitor.visitBlock(steps)

    override fun children(): List<Step> {
        return steps.steps()
    }

    override fun toString(): String {
        return "do"
    }
}
